var/spack/repos/builtin/packages/opencv/package.py

```diff
@@ -2,399 +2,1105 @@
+        "4.1.0",
+        "4.1.1",
+        "4.1.2",
+        "4.2.0",
+        "4.5.0",
+        "4.5.1",
+        "4.5.2",
+        "4.5.4",
+    ]
+    for cv in contrib_vers:
+        resource(
+            name="contrib",
+            git="https://github.com/opencv/opencv_contrib.git",
+            tag="{0}".format(cv),
+            when="@{0}".format(cv),
+        )
+
+    # Patch to fix conflict between CUDA and OpenCV (reproduced with 3.3.0
+    # and 3.4.1) header file that have the same name. Problem is fixed in
+    # the current development branch of OpenCV. See #8461 for more information.
+    patch("dnn_cuda.patch", when="@3.3.0:3.4.1+cuda+dnn")
+
+    patch("opencv3.2_cmake.patch", when="@3.2:3.4.1")
+    patch("opencv3.2_compiler_cmake.patch", when="@3.2")
+    patch("opencv3.2_vtk.patch", when="@3.2+vtk")
+    patch("opencv3.2_regacyvtk.patch", when="@3.2+vtk")
+    patch("opencv3.2_ffmpeg.patch", when="@3.2+videoio")
+    patch("opencv3.2_python3.7.patch", when="@3.2+python3")
+    patch("opencv3.2_fj.patch", when="@3.2 %fj")
+
+    # do not prepend system paths
+    patch("cmake_no-system-paths.patch")
+    patch("clp_cmake.patch")
+    patch("cvv_cmake.patch")
+
+    # OpenCV prebuilt apps (variants)
+    # Defined in `apps/*/CMakeLists.txt` using
+    # `ocv_add_application(...)`
+    apps = [
+        "annotation",
+        "createsamples",
+        "interactive-calibration",
+        "model-diagnostics",
+        "traincascade",
+        "version",
+        "visualisation",
+    ]
+
+    # app variants
+    for app in apps:
+        variant(app, default=False, description="Install {0} app".format(app))
+
+    # app conflicts
+    with when("+annotation"):
+        conflicts("~highgui")
+        conflicts("~imgcodecs")
+        conflicts("~imgproc")
+        conflicts("~videoio")
+
+    with when("+createsamples"):
+        conflicts("~calib3d")
+        conflicts("~features2d")
+        conflicts("~highgui")
+        conflicts("~imgcodecs")
+        conflicts("~imgproc")
+        conflicts("~objdetect")
+        conflicts("~videoio")
+
+    with when("+interactive-calibration"):
+        conflicts("~calib3d")
+        conflicts("~features2d")
+        conflicts("~highgui")
+        conflicts("~imgproc")
+        conflicts("~videoio")
+
+    with when("+model-diagnostics"):
+        conflicts("~dnn")
+
+    with when("+traincascade"):
+        conflicts("~calib3d")
+        conflicts("~features2d")
+        conflicts("~highgui")
+        conflicts("~imgcodecs")
+        conflicts("~imgproc")
+        conflicts("~objdetect")
+
+    with when("+visualisation"):
+        conflicts("~highgui")
+        conflicts("~imgcodecs")
+        conflicts("~imgproc")
+        conflicts("~videoio")
 
     # OpenCV modules (variants)
     # Defined in `modules/*/CMakeLists.txt` using
     # `ocv_add_module(...)` and `ocv_define_module(...)`
     modules = [
-        'apps', 'calib3d', 'core', 'dnn', 'features2d', 'flann', 'gapi', 'highgui',
```

Comment:
If you're going to remove variants, you'll need to update packages that depend on them. For example:
```console
$ grep depends_on.*opencv.*core */package.py
3dtk/package.py:    depends_on('opencv+calib3d+contrib+core+features2d+highgui+imgcodecs+imgproc+ml+videoio', when='+opencv')
caffe/package.py:    depends_on('opencv@3.2.0:3.4.12+core+highgui+imgproc+imgcodecs', when='+opencv')
candle-benchmarks/package.py:    depends_on('opencv@3.2.0: +core +highgui +imgproc +jpeg +png +tiff -dnn ~eigen ~gtk')
dbow2/package.py:    depends_on('opencv+calib3d+core+features2d+highgui+imgproc')
dlib/package.py:    depends_on('opencv+calib3d+core+features2d+highgui+imgproc+imgcodecs')
lbann/package.py:    depends_on('opencv@4.1.0: build_type=RelWithDebInfo +core +highgui '
mivisionx/package.py:    depends_on('opencv@:3.4 +calib3d+core+features2d+highgui+imgcodecs+imgproc+video+videoio+flann+photo+objdetect', type='build')
mxnet/package.py:    depends_on('opencv+core+highgui+imgproc+imgcodecs', when='+opencv')
openvslam/package.py:    depends_on('opencv@3.3.1:+core+imgcodecs+videoio+features2d+calib3d+highgui')
```

Reply:
Yes, that is important. Missed the forest for the trees here. I added a new commit to address this, but the `cntk` package will need more work to get it to build.